numpy/core/src/multiarray/descriptor.c

```diff
@@ -238,6 +238,20 @@ is_datetime_typestr(char const *type, Py_ssize_t len)
     return 0;
 }
 
+// This function checks if val == {'name': any tuple}.
+// See gh-2865, issue #8235 for details
+inline npy_bool _check_if_size_1_and_contains_name_and_tuple(PyObject* val) {
+    npy_bool is_ok = PyDict_Size(val) == 1;
+    if (is_ok) {
+        PyObject* name = PyUnicode_FromString("name");
+        if (!name) return 1;
```

Comment:
Don't silence memoryerror here:
```suggestion
        if (!name) {
            return -1;
        }
```
The caller will need to handle this specially too.